### PR TITLE
receiver: show deploy progress in git push like flynn scale

### LINKED ISSUE #2161: Show deploy progress in git push
We should print the deploy events that happen during the last part of a `git push` like we do in `flynn scale`.

### PR DESCRIPTION
**Pull request by imjorge**
Closes #2161 

```
$ git push flynn master
Counting objects: 40, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (39/39), done.
Writing objects: 100% (40/40), 3.73 KiB | 0 bytes/s, done.
Total 40 (delta 20), reused 0 (delta 0)
-----> Building example...
-----> Node.js app detected

-----> Creating runtime environment

       NPM_CONFIG_LOGLEVEL=error
       NPM_CONFIG_PRODUCTION=true
       NODE_ENV=production
       NODE_MODULES_CACHE=true

-----> Installing binaries
       engines.node (package.json):  0.10.x
       engines.npm (package.json):   2.x

       Resolving node version 0.10.x via semver.io...
       Downloading and installing node 0.10.40...
       Resolving npm version 2.x via semver.io...
       Downloading and installing npm 2.14.13 (replacing version 1.4.28)...

-----> Restoring cache
       Skipping cache (new runtime signature)

-----> Building dependencies
       Pruning any extraneous modules
       Installing node modules (package.json)

-----> Caching build
       Clearing previous node cache
       Saving 1 cacheDirectories (default):
       - node_modules (nothing to cache)

-----> Build succeeded!
       └── (empty)

-----> Discovering process types
       Procfile declares types -> web
       Default process types for Node.js -> web
-----> Compiled slug size is 7.0M
-----> Creating release...
=====> Waiting for web job to start...
-----> new web job is starting
-----> new web job is up
=====> Default web formation scaled to 1
=====> Application deployed
To https://git.dev.localflynn.com/example.git
 * [new branch]      master -> master
 ```

### CODE REVIEW

File: gitreceive/receiver/flynn-receive.go [hunk 1]
```diff
@@ -199,13 +199,10 @@ Options:
...
 		}
-		fmt.Println("=====> Waiting for web job to start...")
 
 		err = watcher.WaitFor(ct.JobEvents{"web": {"up": 1}}, scaleTimeout, func(e *ct.Job) error {
-			switch e.State {
-			case "up":
-				fmt.Println("=====> Default web formation scaled to 1")
-			case "down", "crashed":
+			fmt.Printf("%s ==> %s %s %s\n", time.Now().Format("15:04:05.000"), e.Type, e.ID, e.State)
+			if e.State == "down" || e.State == "crashed" {
```
Comment (titanous) [inline gitreceive/receiver/flynn-receive.go]: This will only print on the first deploy, which is incorrect. I think the timestamp should be at the end in parens and the arrows should be the same length for symmetry.

Comment (imjorge) [inline gitreceive/receiver/flynn-receive.go]: Done e9630d5

File: gitreceive/receiver/flynn-receive.go [hunk 2]
```diff
@@ -176,12 +176,22 @@ Options:
...
+	} else {
+		defer watcher.Close()
+	}
+
 	if err := client.DeployAppRelease(app.Name, release.ID); err != nil {
 		log.Fatalln("Error deploying app release:", err)
 	}
 
-	fmt.Println("=====> Application deployed")
+	fmt.Printf("=====> Application deployed (%s)\n", time.Now().Format("15:04:05.000"))
```
Comment (titanous) [inline gitreceive/receiver/flynn-receive.go]: The application hasn't been deployed yet. `Deploying release` or similar.

File: gitreceive/receiver/flynn-receive.go [hunk 3]
```diff
@@ -189,23 +199,28 @@ Options:
...
-			case "down", "crashed":
+		expected = ct.JobEvents{"web": {"up": 1}}
+	} else {
+		formation, err := client.GetFormation(app.ID, release.ID)
+		if err != nil {
+			log.Fatalln("Error getting formation:", err)
+		}
+		expected = make(ct.JobEvents, len(formation.Processes))
+		for k, v := range formation.Processes {
+			if k == "web" {
```
Comment (titanous) [inline gitreceive/receiver/flynn-receive.go]: I think this will work:

```go
var expected ct.JobEvents
if n, ok := formation.Processes["web"]; ok && n > 0 {
    expected = ct.JobEvents{"web": map[string]int{"up": v}}
}
```

Comment (imjorge) [inline gitreceive/receiver/flynn-receive.go]: done.

File: gitreceive/receiver/flynn-receive.go (hunk 2 above)
Comment (imjorge) [inline gitreceive/receiver/flynn-receive.go]: done.

**Comment by imjorge** [comment]
We are only listening for formation changes for `web` process types. should we be listening for process types with suffix `-web`? And, if so, should we also include them in the default scale?

File: gitreceive/receiver/flynn-receive.go [hunk 4]
```diff
@@ -176,12 +176,22 @@ Options:
...
+	} else {
+		defer watcher.Close()
+	}
+
 	if err := client.DeployAppRelease(app.Name, release.ID); err != nil {
 		log.Fatalln("Error deploying app release:", err)
 	}
 
-	fmt.Println("=====> Application deployed")
+	fmt.Printf("=====> Deploying release (%s)\n", time.Now().Format("15:04:05.000"))
```
Comment (lmars) [inline gitreceive/receiver/flynn-receive.go]: You could put the parens in the time format:

```go
fmt.Println("=====> Deploying release", time.Now().Format("(15:04:05.000)"))
```

File: gitreceive/receiver/flynn-receive.go [hunk 5]
```diff
@@ -189,23 +199,25 @@ Options:
 			Processes: map[string]int{"web": 1},
 		}
 
-		watcher, err := client.WatchJobEvents(app.ID, release.ID)
-		if err != nil {
-			log.Fatalln("Error streaming job events", err)
-			return
-		}
-		defer watcher.Close()
```
Comment (lmars) [inline gitreceive/receiver/flynn-receive.go]: We should continue to only watch for job events here on initial push (in which case the deployment didn't do anything and we need to scale up), and then on subsequent pushes we should print deployment events (which are waited on [here](https://github.com/flynn/flynn/blob/master/controller/client/client.go#L495-L540)).

**Comment by lmars** [comment]
> We are only listening for formation changes for web process types. should we be listening for process types with suffix -web? And, if so, should we also include them in the default scale?

I think it is fine for now to only default scale `web=1`, users can then subsequently scale other process types if need be.

File: gitreceive/receiver/flynn-receive.go (hunk 5 above)
Comment (imjorge) [inline gitreceive/receiver/flynn-receive.go]: This seems different from what @titanous commented earlier.

Comment (lmars) [inline gitreceive/receiver/flynn-receive.go]: Yes, I am superseding his authority on this one :wink:

The deployment waits for job events and emits deployment events, so we should just be logging those deployment events and waiting for either a `failed` or `completed` deployment event (which will also fix #2159).

Comment (imjorge) [inline gitreceive/receiver/flynn-receive.go]: Could you mock for me what is printed in the console for an initial push and subsequent pushes?

Comment (lmars) [inline gitreceive/receiver/flynn-receive.go]: Initial push:

```
-----> Creating release...
=====> Waiting for web job to start...
-----> new web job is starting
-----> new web job is up
=====> Default web formation scaled to 1
=====> Application deployed
```

Subsequent push:

```
-----> Creating release...
-----> new web job is starting
-----> new web job is up
-----> old web job is stopping
-----> old web job is down
=====> Application deployed
```

I have omitted the job IDs because they are not added to deployment events (maybe they should be), and I don't think we need timestamps either.

WDYT?

Comment (imjorge) [inline gitreceive/receiver/flynn-receive.go]: Thinking of adding parameter `logf func(*ct.DeploymentEvent)` to `DeployAppRelease`. This will be `nil` on other call sites and specialized in this case. Also remove the timeout all together?

Comment (lmars) [inline gitreceive/receiver/flynn-receive.go]: Yes the parameter sounds good, leave the timeout I think we should deal with that in a separate PR.

Comment (imjorge) [inline gitreceive/receiver/flynn-receive.go]: Done. Kept the timeout to be address in separate PR.

File: gitreceive/receiver/flynn-receive.go [hunk 6]
```diff
@@ -176,11 +176,20 @@ Options:
 	if err := client.CreateRelease(release); err != nil {
 		log.Fatalln("Error creating release:", err)
 	}
-	if err := client.DeployAppRelease(app.Name, release.ID); err != nil {
-		log.Fatalln("Error deploying app release:", err)
+
+	logf := func(e *ct.DeploymentEvent) {
+		if e.Status == "running" {
```
Comment (lmars) [inline gitreceive/receiver/flynn-receive.go]: Reverse this check and return to reduce the indentation (i.e. `if e.Status != "running" { return }`)

Comment (imjorge) [inline gitreceive/receiver/flynn-receive.go]: Done.

File: controller/client/client.go [hunk 7]
```diff
@@ -492,7 +492,7 @@ func (c *Client) StreamDeployment(d *ct.Deployment, output chan *ct.DeploymentEv
 	}, appEvents)
 }
 
-func (c *Client) DeployAppRelease(appID, releaseID string) error {
+func (c *Client) DeployAppRelease(appID, releaseID string, logf func(*ct.DeploymentEvent)) error {
```
Comment (titanous) [inline controller/client/client.go]: I'm not sold on the function argument, perhaps this function should return an optional `stream.Stream` if a channel is provided? /cc @lmars

Comment (imjorge) [inline controller/client/client.go]: @titanous @lmars any update on how to proceed with this PR?

Comment (lmars) [inline controller/client/client.go]: @titanous I intend to re-work deployments to make them jobs after merging #2744 (which will aid in fixing a bunch of other deployment related issues), so I wonder if we should either 1) go with the current `logf` solution or 2) defer this change to when deployments are re-worked?

Comment (titanous) [inline controller/client/client.go]: I'm fine with deferring.

**Comment by titanous** [comment]
Closing until the deploy system is reworked.
